sound/soc/sof/core.c

```diff
@@ -255,8 +255,9 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 fw_load_err:
 	snd_sof_ipc_free(sdev);
 ipc_err:
-	snd_sof_free_debug(sdev);
 dbg_err:
+	snd_sof_free_debug(sdev);
+dsp_err:
```

Comment:
@lyakh I'd prefer to fix this in debug.c and have a rollback target to call debugfs_remove_recursive() if there are error with some debugfs entries. The error handling is already pretty complex in the probe_continue, so it would be better to follow the policy that a function either succeeds (and allocates resources) or fails (and cleans up any partial allocations). probe_continue() should do the same for the calls it has done, but mixing the two will get us lost.

Reply:
@kv2019i I thought about that too, but I think in this case it isn't needed. The partial success that debug.c might have is that it creates only some of debugfs files and fails to create the rest. But the clean up is the same as in case of a complete success - delete the complete debugfs SOF directory with all successfully created files in it. So, no need to clean up internally.